File: web-app/src/services/loanService.js
import axios from 'axios';

const API_URL = `${process.env.REACT_APP_API_BASE_URL}/loans`;

const authHeader = () => ({
    headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }
});

export const createLoan = async (loanData) => {
    return await axios.post(`${API_URL}/create`, loanData, authHeader());
};

export const getUserLoans = async () => {
    return await axios.get(`${API_URL}/get`, authHeader());
};

export const getAllLoans = async () => {
    return await axios.get(`${API_URL}/all`, authHeader());
};

export const approveLoan = async (loanId) => {
    return await axios.put(`${API_URL}/${loanId}/approve`, {}, authHeader());
};

export const addRepayment = async (loanId, repaymentData) => {
    return await axios.post(`${API_URL}/${loanId}/repayments`, repaymentData, authHeader());
};

File: mini_loan_flutter/lib/services/api/loan_service.dart
import 'package:http/http.dart' as http;
import 'dart:convert';

import 'package:mini_loan_flutter/utils/constants.dart';

class LoanService {

  Future<List<dynamic>> fetchAllLoans(String token) async {
    print("Start fetch");
    final response = await http.get(
      Uri.parse('$baseUrlLoans/all'),
      headers: {'Content-Type': 'application/json', 'Authorization': 'Bearer $token'},
    );

    print(response.body);

    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    } else {
      print(response);
      throw Exception('Failed to fetch loans');
    }
  }

  Future<List<dynamic>> fetchUserLoans(String token) async {
    final response = await http.get(
      Uri.parse('$baseUrlLoans/get'),
      headers: {'Authorization': 'Bearer $token'},
    );

    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    } else {
      throw Exception('Failed to fetch loans');
    }
  }

  Future<void> createLoan(String token, Map<String, dynamic> loanData) async {
    final response = await http.post(
      Uri.parse('$baseUrlLoans/create'),
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $token',
      },
      body: jsonEncode(loanData),
    );

    if (response.statusCode != 201) {
      throw Exception('Failed to create loan');
    }
  }

  Future<void> makeRepayment(String token, String loanId, double amount) async {
    final response = await http.post(
      Uri.parse('$baseUrlLoans/$loanId/repayments'),
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $token',
      },
      body: jsonEncode({'amount': amount}),
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to make repayment');
    }
  }
}
